src/sdformat_urdf.cpp

```diff
@@ -0,0 +1,466 @@
+  }
+  if (1u != sdf_dom.ModelCount()) {
+    errors.emplace_back(
+      sdf::ErrorCode::STRING_READ,
+      "SDFormat xml has multiple models; but only a single model is supported");
+    return nullptr;
+  }
+
+  return convert_model(*sdf_dom.ModelByIndex(0), errors);
+}
+
+urdf::ModelInterfaceSharedPtr
+sdformat_urdf::convert_model(const sdf::Model & sdf_model, sdf::Errors & errors)
+{
+  urdf::ModelInterfaceSharedPtr urdf_model = std::make_shared<urdf::ModelInterface>();
+
+  // copy name
+  urdf_model->name_ = sdf_model.Name();
+
+  // create matching links
+  for (uint64_t l = 0; l < sdf_model.LinkCount(); ++l) {
+    const sdf::Link * sdf_link = sdf_model.LinkByIndex(l);
+
+    if (!sdf_link) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get sdf link");
+      return nullptr;
+    }
+
+    auto pair = urdf_model->links_.emplace(sdf_link->Name(), convert_link(*sdf_link, errors));
+
+    if (!pair.second) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to create element in links map");
+      return nullptr;
+    }
+
+    const std::shared_ptr<urdf::Link> & urdf_link = pair.first->second;
+
+    if (!urdf_link) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to convert sdf::link [" + sdf_link->Name() + "] to urdf::Link");
+      return nullptr;
+    }
+  }
+
+  // sdf canonical link -> urdf root link
+  const sdf::Link * sdf_canonical_link = sdf_model.CanonicalLink();
+  if (!sdf_canonical_link) {
+    errors.emplace_back(
+      sdf::ErrorCode::STRING_READ,
+      "Failed to get sdf canonical link");
+    return nullptr;
+  }
+  auto iter = urdf_model->links_.find(sdf_canonical_link->Name());
+  if (iter == urdf_model->links_.end()) {
+    errors.emplace_back(
+      sdf::ErrorCode::STRING_READ,
+      "Failed to find sdf canonical link [" + sdf_canonical_link->Name() + "]");
+    return nullptr;
+  }
+  urdf_model->root_link_ = iter->second;
+
+  // create matching joints
+  for (uint64_t j = 0; j < sdf_model.JointCount(); ++j) {
+    const sdf::Joint * sdf_joint = sdf_model.JointByIndex(j);
+
+    if (!sdf_joint) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get sdf joint");
+      return nullptr;
+    }
+
+    auto pair = urdf_model->joints_.emplace(sdf_joint->Name(), convert_joint(*sdf_joint, errors));
+
+    if (!pair.second) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to create element in joints map");
+      return nullptr;
+    }
+
+    const std::shared_ptr<urdf::Joint> & urdf_joint = pair.first->second;
+    if (!urdf_joint) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to convert sdf::joint [" + sdf_joint->Name() + "] to urdf::Joint");
+      return nullptr;
+    }
+  }
+
+  // Start with root link (sdf canonical frame) and recurse this function:
+  //  
+  // TODO what about coordinate frames?
```

Comment:
Er, I'm not sure if I understand this TODO. Why are coordinate frames important? I think you just resolve the pose and you're good (which is what you're doing below)?

Reply:
I forgot to delete the TODO after adding the code below

Reply:
Deleted TODO in 4f0ef012274e6f79d8c38010966a096feb9c26a7